File: tests_c/pthread.c
#include "util.h"

typedef int   (*pthread_create_fn)(sd_u64*, void*, void* (*)(void*), void*);
typedef int   (*pthread_join_fn)(sd_u64, void**);
typedef int*  (*errno_location_fn)(void);
typedef int   (*open_fn)(const char*, int);
typedef void* (*malloc_fn)(sd_u64);
typedef void  (*free_fn)(void*);

_Thread_local sd_u32 tl_u32 = 5;
_Thread_local sd_u64 tl_u64[4] = {0xc0ffee, 0xc0ffee1, 0xc0ffee2, 0xc0ffee3};

static errno_location_fn g_errno;
static open_fn           g_open;
static malloc_fn         g_malloc;
static free_fn           g_free;
static sd_u64            main_tp;

static void* body(void* arg) {
  (void)arg;

  sd_u64 tp = get_thread_pointer();
  ASSERT(tp != 0);
  ASSERT(tp != main_tp);

  ASSERT_EQ(tl_u32, 5);
  ASSERT_EQ(tl_u64[0], 0xc0ffee);
  ASSERT_EQ(tl_u64[3], 0xc0ffee3);
  tl_u32 = 100;

  int* e = g_errno();
  ASSERT(e != 0);
  g_open("blah", 0);
  ASSERT_EQ(*e, 2) // ENOENT

  char* p = (char*)g_malloc(4096);
  ASSERT(p != 0);
  p[4095] = 1;
  ASSERT_EQ(p[4095], 1);
  g_free(p);

  return 0;
}

int main(int argc, char** argv) {
  (void)argc;
  (void)argv;

  ASSERT(sd_got.success != 0);

  void* libc = sd_got.dlopen("libc.so.6", SD_RTLD_NOW);
  ASSERT(libc != 0);

  pthread_create_fn create = (pthread_create_fn)sd_got.dlsym(libc, "pthread_create");
  pthread_join_fn   join   = (pthread_join_fn)sd_got.dlsym(libc, "pthread_join");
  g_errno  = (errno_location_fn)sd_got.dlsym(libc, "__errno_location");
  g_open   = (open_fn)          sd_got.dlsym(libc, "open");
  g_malloc = (malloc_fn)        sd_got.dlsym(libc, "malloc");
  g_free   = (free_fn)          sd_got.dlsym(libc, "free");
  ASSERT(create   != 0);
  ASSERT(join     != 0);
  ASSERT(g_errno  != 0);
  ASSERT(g_open   != 0);
  ASSERT(g_malloc != 0);
  ASSERT(g_free   != 0);

  main_tp = get_thread_pointer();
  tl_u32 = 7;

  sd_u64 tid = 0;
  ASSERT_EQ(create(&tid, 0, body, 0), 0);
  ASSERT_EQ(join(tid, 0), 0);

  ASSERT_EQ(tl_u32, 7);

  return 0;
}

File: tests_zig/pthread.zig
const std = @import("std");
const sd = @import("static_dynamic");
const utils = @import("util.zig");

pub const _start = {};

threadlocal var tl_u32: u32 = 5;
threadlocal var tl_arr: [4]u64 = .{ 0xc0ffee, 0xc0ffee1, 0xc0ffee2, 0xc0ffee3 };

var libc: utils.Libc = undefined;
var main_tp: usize = 0;
var area_size: usize = 0;
var counter: std.atomic.Value(u32) = .init(0);

fn bump() void {
    _ = counter.fetchAdd(1, .monotonic);
}

fn body(_: ?*anyopaque) callconv(.c) ?*anyopaque {
    const tp = utils.get_thread_pointer();
    utils.assert(tp != 0);
    utils.assert(tp != main_tp);

    utils.assert(std.os.linux.elf_aux_maybe != null);
    utils.assert_eq(std.os.linux.getauxval(std.elf.AT_PAGESZ), 4096);
    utils.assert_eq(std.os.linux.tls.area_desc.size, area_size);

    utils.assert_eq(tl_u32, 5);
    utils.assert_eq(tl_arr[0], 0xc0ffee);
    utils.assert_eq(tl_arr[3], 0xc0ffee3);
    tl_u32 = 100;

    const e = libc.errno_location();
    utils.assert(@as(u64, @intFromPtr(e)) != 0);
    _ = libc.open("blah", 0);
    utils.assert_eq(e.*, 2);

    const p = libc.malloc(4096);
    utils.assert(p != null);
    @as([*]u8, @ptrCast(p.?))[4095] = 1;
    utils.assert_eq(@as([*]u8, @ptrCast(p.?))[4095], 1);
    libc.free(p);

    const page = std.heap.page_allocator;
    const a = page.alloc(u8, 8192) catch unreachable;
    a[8191] = 1;
    utils.assert_eq(a[8191], 1);
    page.free(a);

    const smp = std.heap.smp_allocator;
    const b = smp.alloc(u8, 5000) catch unreachable;
    b[4999] = 2;
    utils.assert_eq(b[4999], 2);
    smp.free(b);

    var dbg: std.heap.DebugAllocator(.{}) = .init;
    const c = dbg.allocator().alloc(u32, 64) catch unreachable;
    dbg.allocator().free(c);

    const th = std.Thread.spawn(.{}, bump, .{}) catch unreachable;
    th.join();
    utils.assert_eq(counter.load(.monotonic), 1);

    utils.assert_eq(std.Thread.getCurrentId(), @as(u32, @intCast(std.os.linux.gettid())));

    return null;
}

export fn main(argc: u64, argv: [*]const [*:0]const u8) callconv(.c) i32 {
    sd.init(argc, argv);
    utils.assert(sd.got.result.success != 0);
    libc = utils.Libc.load(sd.got);

    main_tp = utils.get_thread_pointer();
    area_size = std.os.linux.tls.area_desc.size;
    tl_u32 = 7;

    var tid: usize = 0;
    utils.assert_eq(libc.pthread_create(&tid, null, body, null), 0);
    utils.assert_eq(libc.pthread_join(tid, null), 0);

    utils.assert_eq(tl_u32, 7);

    return 0;
}
